### PR TITLE
added BindingWithGrammar.md to document vscode-xml grammar binding

### LINKED ISSUE #567: Add documentation page for binding to grammar/schema
As per comment https://github.com/redhat-developer/vscode-xml/pull/565#discussion_r678572552, we should document the functionality of the binding wizard with it's own documentation page to detail this notable vscode-xml feature (and to reference elsewhere).

### PR DESCRIPTION
**Pull request by AlexXuChen**
Added BindingWithGrammar.md to document vscode-xml grammar binding using existing grammar with binding wizard and how to generate a new grammar file.

Closes #567 

Signed-off-by: Alexander Chen alchen@redhat.com

### CODE REVIEW

File: docs/Features/XMLFeatures.md [hunk 1]
```diff
@@ -128,25 +128,7 @@ See [XML Validation](../Validation.md#xml-validation) for more details.
...
-### Generate DTD from XML
-
-Using an XML file, a Document Type Definition/DTD file can be generated from the opening tag in the XML file. With the cursor on the first opening tag, use `Ctrl + .` or `Quick Fix...` and select `Generate foo.dtd and bind with *` to create the file in the the same directory.
-
-![GenerateDTDFromXML](../images/Features/GenerateDTDFromXML.gif)
-
-#### Read more
-
-See [Validation with DTD Grammar](../Validation.md#validation-with-dtd-grammar) for more.
+See [Binding with New Grammar](../BindingWithGrammar.md#quick-fix/codeaction) for info on grammar generation.
```
Comment (datho7561) [inline docs/Features/XMLFeatures.md:131]: This link doesn't appear to work in the built-in documentation. I get sent to the end of the "Binding With Grammar" page instead of the Quick Fix/ Code Action section

File: docs/Features/XMLFeatures.md [hunk 2]
```diff
@@ -128,25 +128,7 @@ See [XML Validation](../Validation.md#xml-validation) for more details.
...
```
Comment (datho7561) [inline docs/Features/XMLFeatures.md:132]: It would be nice to add another section to this document to list that you can bind to an existing grammar as well. This would serve to make `XMLFeatures.md` into a mostly complete feature list of vscode-xml. For the content, you can just link to the "Binding With Grammar" page like you did with "Generate Grammar".

File: docs/BindingWithGrammar.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    <xs:complexType>
+      <xs:sequence>
+        <xs:element name="bar" />
+      </xs:sequence>
+    </xs:complexType>
+  </xs:element>
+</xs:schema>
+```
+
+This file (or a corresponding `.dtd` file) can be bound using a [binding wizard](#the-xml-binding-wizard), which can be triggered by [Command](#command), [CodeLens](#codelens) or [Quick Fix/CodeAction](#quick-fix/codeaction).
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:34]: ... using *the* binding wizard ...

Comment (datho7561) [inline docs/BindingWithGrammar.md:34]: The link to "Quick Fix/CodeAction" here also doesn't work in VS Code

File: docs/BindingWithGrammar.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+### The XML Binding Wizard
+
+After opening the binding wizard using one of the methods mentioned above, a dropdown with the following 2 options will appear at the Command Palette:
+
+![Binding Wizard Dropdown](./images/BindingWithGrammar/BindingWizardDropdown.png)
+
+1. Standard (xsi, DOCTYPE)
+
+This option will bind the grammar file by adding the `xmlns:xsi`(with the value as "http://www.w3.org/2001/XMLSchema-instance") and `xsi:noNamespaceSchemaLocation`(with the value as the path to the grammar file) attributes to the root tag:
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:44]: Add a space after both of the ` characters

File: docs/BindingWithGrammar.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+![Bind To Grammar Command](./images/BindingWithGrammar/BindToGrammarCommand.png)
+
+If the current XML document is not bound using an `xml-model` tag or a root element doesn't have a `xsi:noNameSpaceLocation` attribute, the command will have the user select a existing `.xsd/.dtd` grammar file to bind to the XML document.
+
+Otherwise, an error will be thrown on the client if the XML document already has a bound grammar/schema.
+
+### CodeLens
+
+When CodeLens is enabled, an unbound XML document will display a CodeLens above the root element as follows:
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:78]: ... When CodeLenses are enabled ...

I think it might be worthwhile to link to the setting to enable code lens in `Preferences.md#code-lens`

File: docs/BindingWithGrammar.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+![Bind To Grammar Command](./images/BindingWithGrammar/BindToGrammarCodeLens.png)
+
+### Quick Fix/CodeAction
+
+For any unbound XML document, a Quick Fix suggestion will appear beside the root element.
+
+![Bind To Grammar CodeAction](./images/BindingWithGrammar/BindToGrammarCodeAction.png)
+
+With the cursor on the first opening tag, use `Ctrl + .` or `Quick Fix...` and select the "Open binding wizard to bind existing grammar/schema"
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:88]: You can also click on the lightbulb

File: docs/BindingWithGrammar.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+![Bind To Grammar CodeAction Dropdown](./images/BindingWithGrammar/BindToGrammarCodeActionDropdown.png)
+
+## Binding with New Grammar
+
+Consider the case where the directory only contains `foo.xml` (i.e. there does not exist an appropriate `.xsd/.dtd` file to bind).
+
+A file can be generated using a [Quick Fix/CodeAction](#quick-fix/codeaction-1).
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:96]: This link also doesn't work in VS Code

File: docs/BindingWithGrammar.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+### Quick Fix/CodeAction
+
+#### Generate XSD from XML
+
+When an unbound XML file is open, an XML Schema/XSD file can be generated from the opening tag in the XML file.
+
+With the cursor on the first opening tag, use `Ctrl + .` or `Quick Fix...` and select "Generate foo.xsd and bind with *" to create the file in the the same directory.
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:104]: See note above about lightbulb

File: docs/BindingWithGrammar.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+![GenerateXSDFromXML](./images/BindingWithGrammar/GenerateXSDFromXML.gif)
+
+#### Generate DTD from XML
+
+When an unbound XML file is open, a Document Type Definition/DTD file can be generated from the opening tag in the XML file.
+
+With the cursor on the first opening tag, use `Ctrl + .` or `Quick Fix...` and select "Generate foo.dtd and bind with *" to create the file in the the same directory.
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:112]: See note above about lightbulb

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: docs/Features/XMLFeatures.md (hunk 2 above)
Comment (AlexXuChen) [reply docs/Features/XMLFeatures.md:136]: I generalized to a header called "Binding Grammar to an XML Document"

File: docs/BindingWithGrammar.md (hunk 3 above)
Comment (AlexXuChen) [reply docs/BindingWithGrammar.md]: The link seems to work without a change, could you confirm?

File: docs/BindingWithGrammar.md (hunk 7 above)
Comment (AlexXuChen) [reply docs/BindingWithGrammar.md:96]: Same as above

File: docs/BindingWithGrammar.md (hunk 4 above)
Comment (datho7561) [reply docs/BindingWithGrammar.md]: @AlexXuChen other "after", i.e. between the ` and the (. My bad I should have explained this clearer

File: docs/BindingWithGrammar.md (hunk 3 above)
Comment (datho7561) [reply docs/BindingWithGrammar.md]: It doesn't seem to work for me on Linux or windows on the latest stable VS Code. I think maybe rewording the title to avoid the `/` should work?

File: docs/BindingWithGrammar.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+### The XML Binding Wizard
+
+After opening the binding wizard using one of the methods mentioned above, a dropdown with the following 2 options will appear at the Command Palette:
+
+![Binding Wizard Dropdown](./images/BindingWithGrammar/BindingWizardDropdown.png)
+
+1. Standard (xsi, DOCTYPE)
+
+This option will bind the grammar file by adding the `xmlns:xsi`v(with the value as "http://www.w3.org/2001/XMLSchema-instance") and `xsi:noNamespaceSchemaLocation` (with the value as the path to the grammar file) attributes to the root tag:
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:44]: There's a `v` between the ` and the (

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: docs/Features/XMLFeatures.md [hunk 11]
```diff
@@ -120,33 +120,19 @@ LemMinX will show syntax errors in your XML documentation, and will provide quic
...
 
 ### Read more
 
 See [XML Validation](../Validation.md#xml-validation) for more details.
 
-## Generate Grammar from XML
+## Binding Grammar to an XML Document
 
-### Generate XSD from XML
+For any XML document, should you need to bind a grammar or schema with the file, you can do by either using an existing grammar file, or by generating one using the built-in schema generator.
```
Comment (AlexXuChen) [inline docs/Features/XMLFeatures.md:131]: "...you can do _so_ by..."

File: docs/BindingWithGrammar.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+### The XML Binding Wizard
+
+After opening the binding wizard using one of the methods mentioned above, a dropdown with the following 2 options will appear at the Command Palette:
+
+![Binding Wizard Dropdown](./images/BindingWithGrammar/BindingWizardDropdown.png)
+
+1. Standard (xsi, DOCTYPE)
+
+This option will bind the grammar file by adding the `xmlns:xsi` (with the value as "http://www.w3.org/2001/XMLSchema-instance") and `xsi:noNamespaceSchemaLocation` (with the value as the path to the grammar file) attributes to the root tag:
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:44]: Mention that for DOCTYPE its different.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: docs/BindingWithGrammar.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+If the selected file is an XSD file, this option will bind the grammar file by adding the `xmlns:xsi` (with the value as "http://www.w3.org/2001/XMLSchema-instance") and `xsi:noNamespaceSchemaLocation` (with the value as the path to the grammar file) attributes to the root tag:
+
+```xml
+<foo xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+ xsi:noNamespaceSchemaLocation="foo.xsd">
+    <bar />
+</foo>
+```
+
+If the selected file is an XSD file, this option will bind the grammar file by adding a `<!DOCTYPE <root-tag-name> SYSTEM "<grammar-file-path>">` tag at the beginning of the document:
```
Comment (datho7561) [inline docs/BindingWithGrammar.md:59]: ...is a DTD file...

**Review by datho7561 — CHANGES_REQUESTED** [review]

**Review by datho7561 — APPROVED** [review]
LGTM

File: docs/BindingWithGrammar.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+### The XML Binding Wizard
+
+After opening the binding wizard using one of the methods mentioned above, a dropdown with the following 2 options will appear at the Command Palette:
+
+![Binding Wizard Dropdown](./images/BindingWithGrammar/BindingWizardDropdown.png)
+
+1. Standard (xsi, DOCTYPE)
+
+If the selected file is an XSD file, this option will bind the grammar file by adding the `xmlns:xsi` (with the value as "http://www.w3.org/2001/XMLSchema-instance") and `xsi:noNamespaceSchemaLocation` (with the value as the path to the grammar file) attributes to the root tag:
```
Comment (angelozerr) [inline docs/BindingWithGrammar.md:50]: xsi:schemaLocation is supported two. When you try to bind a xsd which have a namespace, so please add that.

File: docs/BindingWithGrammar.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+```
+
+and a DTD document, `foo.dtd` (in the same directory as `foo.xml`), defined as follows:
+```dtd
+<!ELEMENT foo (bar)>
+<!ELEMENT bar EMPTY>
+```
+
+Either file can be bound using the [XML Binding Wizard](#the-xml-binding-wizard), which can be triggered by [Command](#command), [CodeLens](#codelens) or [Quick Fix](#quick-fix).
+
```
Comment (angelozerr) [inline docs/BindingWithGrammar.md:41]: I think you should add a sample demo with CodeLens here to see in action the process and see that you can select xsd file.

File: docs/BindingWithGrammar.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+After selecting an option, you can select the desired grammar file using your file explorer.
+
+### Command
+
+The command "XML: Bind to grammar/schema file" can be executed from the VSCode command palette, as well as be bound to a shortcut.
+
+![Bind To Grammar Command](./images/BindingWithGrammar/BindToGrammarCommand.png)
+
+If the current XML document is not bound using an `xml-model` tag or a root element doesn't have a `xsi:noNameSpaceLocation` attribute, the command will have the user select a existing `.xsd/.dtd` grammar file to bind to the XML document.
```
Comment (angelozerr) [inline docs/BindingWithGrammar.md:89]: It's more complex than that, because xml can be boundwith a xsd by using XML catalog just by declaring a namespace. You can bind an xml with file associations too https://github.com/redhat-developer/vscode-xml/blob/master/docs/Preferences.md#file-associations and in this case you don't touch the xml.

File: docs/BindingWithGrammar.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+With the cursor on the first opening tag, use `Ctrl + .`, `Quick Fix...` or the lightbulb that appears and select the "Open binding wizard to bind existing grammar/schema"
+
+![Bind To Grammar CodeAction Dropdown](./images/BindingWithGrammar/BindToGrammarCodeActionDropdown.png)
+
+## Binding with New Grammar
+
+Consider the case where the directory only contains `foo.xml` (i.e. there does not exist an appropriate `.xsd/.dtd` file to bind).
+
+A file can be generated using a [Quick Fix](#quick-fix-1).
+
```
Comment (angelozerr) [inline docs/BindingWithGrammar.md:114]: I will add a demo here

File: docs/BindingWithGrammar.md [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+### Quick Fix
+
+For any unbound XML document, a Quick Fix suggestion will appear beside the root element.
+
+![Bind To Grammar CodeAction](./images/BindingWithGrammar/BindToGrammarCodeAction.png)
+
+With the cursor on the first opening tag, use `Ctrl + .`, `Quick Fix...` or the lightbulb that appears and select the "Open binding wizard to bind existing grammar/schema"
+
+![Bind To Grammar CodeAction Dropdown](./images/BindingWithGrammar/BindToGrammarCodeActionDropdown.png)
```
Comment (angelozerr) [inline docs/BindingWithGrammar.md:107]: I think the image is out of dated because the quick fix title is bad.

File: docs/BindingWithGrammar.md (hunk 17 above)
Comment (AlexXuChen) [reply docs/BindingWithGrammar.md:114]: In my next push, I added a binding wizard demo, so I believe that should work

File: docs/BindingWithGrammar.md (hunk 18 above)
Comment (AlexXuChen) [reply docs/BindingWithGrammar.md:107]: Could you pinpoint what part of the quick fix title is bad?

File: docs/BindingWithGrammar.md [hunk 19]
```diff
@@ -0,0 +1,138 @@
+# Binding with Grammar
+
+[vscode-xml](https://github.com/redhat-developer/vscode-xml) provides support for automatic XML schema/grammar binding through generation and/or using an existing `.xsd/.dtd` file.
+
```
Comment (angelozerr) [inline docs/BindingWithGrammar.md:4]: Here I will show directly what looks like the support mean. In other words, I will add:

Here a demo which show how to bind an XML file with an existing XSD file. (and you add a demo with Codelens which open a xsd and bind it).

File: docs/BindingWithGrammar.md (hunk 18 above)
Comment (angelozerr) [reply docs/BindingWithGrammar.md:114]: image

**Comment by angelozerr** [comment]
Great documentation @AlexXuChen, thanks!
